src/diamond/collector.py

```diff
@@ -56,8 +56,14 @@ def get_hostname(config, method=None):
                                     stdout=subprocess.PIPE)
             hostname = proc.communicate()[0].strip()
             if proc.returncode != 0:
-                raise subprocess.CalledProcessError(proc.returncode,
-                                                    config['hostname'])
+                skip_errors = config.get('hostname_cache_skip_errors')
+                # If we aren't explicitly skipping errors OR we have no cached
+                # result, raise an exception.
+                if not skip_errors or method not in get_hostname.cached_results:
+                    raise subprocess.CalledProcessError(proc.returncode,
+                                                        config['hostname'])
+                else:  # Recycle the previously cached version.
+                    hostname = get_hostname.cached_results[method][0]
```

Comment:
Yeah, should definitely log it.

Unfortunately there's no introspective metric publication mechanism, so it wouldn't be straightforward to self-publish a warning.